### PR TITLE
Picturae import

### PR DESCRIPTION
**Pull request by delarocamcas**
Files changed:
metadata_tools.py:
    added functions, for exif read, iptc read, exif attach, iptc attach , exif code to tag, and remove_exif_tags (still in draft).

test_metadata_tools.py: unittests for exif read, iptc read, exif attach, iptc attach , file is larger than. and convert code to tag

### CODE REVIEW

File: metadata_tools.py [hunk 1]
```diff
@@ -39,173 +29,126 @@ def is_file_larger_than(self, filepath: str, size_in_mb) -> bool:
...
+    #             image_path
+    #
+    #         """
+    #     with ExifTool() as et:
+    #         et.execute(f"-IPTC:{','.join(fields_to_remove)}=")
+    #         et.execute(f"-IPTC:{','.join([f'iptc:{field}' for field in fields_to_remove])}=")
+    #
+    #     print("IPTC fields removed successfully.")
+
+    def iptc_attach_metadata(self, iptc_dict: dict, path: str):
```
Comment (delarocamcas) [inline metadata_tools.py:49]: attach metadata with iptcinfo3

File: metadata_tools.py [hunk 2]
```diff
@@ -39,173 +29,126 @@ def is_file_larger_than(self, filepath: str, size_in_mb) -> bool:
...
+                self.logger.error(f"Error saving IPTC metadata: {str(e)}")
+        else:
+            raise ValueError("None Returned")
+
+    def read_iptc_metadata(self, path: str):
+        """reads iptc metadata of image and returns the dictionary"""
+        info = IPTCInfo(path, force=True)
+        return info
+
+    def attach_exif_metadata(self, path, exif_dict):
```
Comment (delarocamcas) [inline metadata_tools.py:68]: attach exif metadata using subprocesses and ExifTools

File: metadata_tools.py [hunk 3]
```diff
@@ -39,173 +29,126 @@ def is_file_larger_than(self, filepath: str, size_in_mb) -> bool:
...
+
+        for exif_code, exif_value in exif_dict.items():
+            exif_code = self.exif_code_to_tag(exif_code)
+            command = ['exiftool', f"-{exif_code}={exif_value}", path]
+            subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        print("exif added succesfully")
+
+
+    # still doesn't work yet, but close
+    def remove_exif_tags(self, path, exif_codes):
```
Comment (delarocamcas) [inline metadata_tools.py:82]: draft function, not fully functional

File: metadata_tools.py [hunk 4]
```diff
@@ -39,173 +29,126 @@ def is_file_larger_than(self, filepath: str, size_in_mb) -> bool:
...
+                if code in exif_data:
+                    exif_data[code] = None
+
+            self.attach_exif_metadata(path=path, exif_dict=exif_data)
+
+            print("EXIF tags removed successfully")
+        except Exception as e:
+            print(f"Error removing EXIF tags: {e}")
+
+    def exif_code_to_tag(self, exif_code):
```
Comment (delarocamcas) [inline metadata_tools.py:103]: converts exif code number to tag name in string format

File: metadata_tools.py [hunk 5]
```diff
@@ -39,173 +29,126 @@ def is_file_larger_than(self, filepath: str, size_in_mb) -> bool:
         # Convert the size to megabytes
         size_in_mb_actual = size_in_bytes / (1024 * 1024)
 
+        print(size_in_mb_actual)
+
```
Comment (foozleface) [inline metadata_tools.py:33]: make debug

File: metadata_tools.py [hunk 6]
```diff
@@ -39,173 +29,126 @@ def is_file_larger_than(self, filepath: str, size_in_mb) -> bool:
...
-            # Convert the IPTC tag and value to bytes
-            iptc_tag_bytes = bytes(iptc_tag, encoding="utf-8")
-            iptc_value_bytes = bytes(iptc_value, encoding="utf-8")
-
-            # Calculate the length of the IPTC tag and value
-            iptc_tag_length = len(iptc_tag_bytes)
-            iptc_value_length = len(iptc_value_bytes)
-
-            # Write the IPTC tag and value to the image data
-            image_data.extend(b"\x1C" + iptc_tag_bytes + b"\x1E" + iptc_value_bytes)
```
Comment (delarocamcas) [inline metadata_tools.py:134]: removed redundant functions

File: metadata_tools.py [hunk 7]
```diff
@@ -39,173 +29,126 @@ def is_file_larger_than(self, filepath: str, size_in_mb) -> bool:
         # Convert the size to megabytes
         size_in_mb_actual = size_in_bytes / (1024 * 1024)
 
+        print(size_in_mb_actual)
+
         # Compare the actual size with the specified size
         return size_in_mb_actual > size_in_mb
```
Comment (foozleface) [inline metadata_tools.py:36]: if your'e leaving a block of commented code in place, put a header about why we're keepoing it. Or trash it.

File: metadata_tools.py [hunk 8]
```diff
@@ -39,173 +29,126 @@ def is_file_larger_than(self, filepath: str, size_in_mb) -> bool:
...
 
-    # print(md)
+if __name__ == "__main__":
+    # print("Running tests...")
+    #
+    md = MetadataTools()
+
+    # # test code commented out
+    #
+    # md.iptc_attach_metadata(path='tests/test_images/test_image.jpg', iptc_dict={
```
Comment (delarocamcas) [inline metadata_tools.py:137]: I probably should remove this test run code, not sure if I want to leave it around. maybe redundant due to unittests already existing

File: tests/test_images/hexdump.sh [hunk 9]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+xxd -p test_image.jpg > test_hexdump.txt
```
Comment (delarocamcas) [inline tests/test_images/hexdump.sh:2]: hex dump ssh, for command line

File: metadata_tools.py [hunk 10]
```diff
@@ -39,173 +29,126 @@ def is_file_larger_than(self, filepath: str, size_in_mb) -> bool:
...
+            exif_dict: dictionary of exif terms using exif codes, and new values to assign"""
+
+        for exif_code, exif_value in exif_dict.items():
+            exif_code = self.exif_code_to_tag(exif_code)
+            command = ['exiftool', f"-{exif_code}={exif_value}", path]
+            subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        print("exif added succesfully")
+
+
+    # still doesn't work yet, but close
```
Comment (foozleface) [inline metadata_tools.py:81]: still true? maybe comment this out if so?

**Review by foozleface — CHANGES_REQUESTED** [review]
Overall - let's make this a class that has a bit of state. Make the interface a kind of for-dummies version. I want a class that on init takes an image path. e,g,: metadata_editor = MetadataEditor('foo/bar/baz.jpg'). I then want interfaces with "add exif data" or "add iptc data" that take not a dict but a single pair - the payload, and the code. If we're confident of the codes, hardcode them as consts in the class. e.g.: "self.COPYRIGHT_EXIF = 1234"
and then, finally, "medatadathingy.save()"

File: metadata_tools.py [hunk 11]
```diff
@@ -1,211 +1,131 @@
...
+    #         print("EXIF tags removed successfully")
+    #     except Exception as e:
+    #         print(f"Error removing EXIF tags: {e}")
+
+    def exif_code_to_tag(self, exif_code):
+        """converts exif code into the string of the tag name
+            args:
+                exif_code: the integer code of an exif tag to convert to TAG"""
+        tag_name = exifread.tags.EXIF_TAGS.get(exif_code, "Unknown Tag")
+        return tag_name[0]
```
Comment (foozleface) [inline metadata_tools.py:106]: Raise an exception here I think

**Review by foozleface — APPROVED** [review]
Add the errror raise and commit at will

**Pull request merged**